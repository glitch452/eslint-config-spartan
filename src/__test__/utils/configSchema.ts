import { z } from 'zod';

const INITIAL_ECMA_VERSION = 3;
const INITIAL_ECMA_YEAR = 2015;
const NEXT_YEAR = new Date().getFullYear() + 1;
const MIN_ECMA_VERSION = 5;
const MAX_ECMA_VERSION = NEXT_YEAR - INITIAL_ECMA_YEAR + MIN_ECMA_VERSION + 1;

const ecmaVersionSchema = z.union([
  z.literal('latest'),
  z.literal(INITIAL_ECMA_VERSION),
  z.number().int().min(MIN_ECMA_VERSION).max(MAX_ECMA_VERSION),
  z.number().int().min(INITIAL_ECMA_YEAR).max(NEXT_YEAR),
]);
const sourceTypeSchema = z.enum(['script', 'module', 'commonjs']);
const globalConfSchema = z.boolean().or(z.enum(['off', 'readable', 'readonly', 'writable', 'writeable']));

const MIN_SEVERITY = 0;
const MAX_SEVERITY = 2;
const severitySchema = z.number().int().min(MIN_SEVERITY).max(MAX_SEVERITY);
const stringSeveritySchema = z.enum(['off', 'warn', 'error']);

const objectMetaPropertiesSchema = z.object({
  /** @deprecated Use `meta.name` instead. */
  name: z.string().optional(),
  /** @deprecated Use `meta.version` instead. */
  version: z.string().optional(),
  meta: z.object({ name: z.string().optional(), version: z.string().optional() }).optional(),
});

const esTreeParserSchema = z.union([
  objectMetaPropertiesSchema.extend({ parse: z.function() }),
  objectMetaPropertiesSchema.extend({ parseForESLint: z.function() }),
]);

const processorSchema = objectMetaPropertiesSchema.extend({
  supportsAutofix: z.boolean().optional(),
  preprocess: z.function(),
  postprocess: z.function(),
});

const parserOptionsSchema = z
  .object({
    ecmaVersion: ecmaVersionSchema,
    sourceType: sourceTypeSchema,
    ecmaFeatures: z
      .object({
        globalReturn: z.boolean(),
        impliedStrict: z.boolean(),
        jsx: z.boolean(),
        experimentalObjectRestSpread: z.boolean(),
      })
      .partial()
      .catchall(z.any()),
  })
  .partial()
  .catchall(z.any());

const languageOptionsSchema = z
  .object({
    ecmaVersion: ecmaVersionSchema,
    sourceType: sourceTypeSchema,
    globals: z.record(globalConfSchema),
    parser: esTreeParserSchema,
    parserOptions: parserOptionsSchema,
  })
  .partial();

const linterOptionsSchema = z
  .object({
    noInlineConfig: z.boolean().optional(),
    reportUnusedDisableDirectives: z.union([severitySchema, stringSeveritySchema, z.boolean()]).optional(),
  })
  .optional();

const externalSpecifierSchema = z.object({
  name: z.string().optional(),
  url: z.string().optional(),
});

const replacedBySchema = z.object({
  message: z.string().optional(),
  url: z.string().optional(),
  plugin: externalSpecifierSchema.optional(),
  rule: externalSpecifierSchema.optional(),
});

const deprecatedInfoSchema = z.object({
  message: z.string().optional(),
  url: z.string().optional(),
  replacedBy: replacedBySchema.array().optional(),
  deprecatedSince: z.string().optional(),
  availableUntil: z.string().nullable().optional(),
});

const ruleMetaDataSchema = z
  .object({
    docs: z
      .object({
        description: z.string(),
        category: z.string(),
        // Added z.string() to accommodate storybook plugin
        // Added z.record(z.unknown()) to accommodate typescript-eslint plugin
        recommended: z.union([z.boolean(), z.string(), z.record(z.unknown())]),
        url: z.string(),
      })
      .partial(),
    messages: z.record(z.string()),
    // Added z.null() added to accommodate the cypress plugin
    fixable: z.enum(['code', 'whitespace']).nullable(),
    schema: z.unknown(),
    deprecated: z.boolean().or(deprecatedInfoSchema),
    replacedBy: z.string().array().readonly(),
    // Added 'error' to accommodate the security plugin
    type: z.enum(['problem', 'suggestion', 'layout', 'error']),
    hasSuggestions: z.boolean(),
  })
  .partial();

const ruleModuleSchema = z.object({
  create: z.function(),
  meta: ruleMetaDataSchema.optional(),
});

const pluginSchema = z
  .object({
    configs: z.record(z.unknown()),
    environments: z.record(
      z.object({ globals: z.record(globalConfSchema), parserOptions: parserOptionsSchema }).partial(),
    ),
    processors: z.record(processorSchema),
    rules: z.record(ruleModuleSchema),
  })
  .partial();

export const configSchema = z
  .object({
    name: z.string(),
    files: z.array(z.string().or(z.string().array())),
    ignores: z.string().array(),
    languageOptions: languageOptionsSchema,
    linterOptions: linterOptionsSchema,
    processor: z.string().or(processorSchema),
    plugins: z.record(pluginSchema),
    rules: z.unknown(),
    settings: z.record(z.unknown()),
  })
  .partial();
