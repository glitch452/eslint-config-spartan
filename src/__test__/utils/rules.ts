import { Linter, Rule } from 'eslint';
import { isOff, isWarn } from '../../utils/index.js';

/**
 * Extract a list of rule names that are 'enabled' i.e. set to 'warn' or 'error'
 * @param rules
 */
export function getEnabledRules(rules: Partial<Record<string, Linter.RuleEntry>> = {}): string[] {
  return Object.entries(rules)
    .filter(([, entry]) => entry && !isOff(entry))
    .map(([name]) => name);
}

/**
 * Extract a list of rule names that are set to 'warn'
 * @param rules
 */
export function getWarnRules(rules: Partial<Record<string, Linter.RuleEntry>> = {}): string[] {
  return Object.entries(rules)
    .filter(([, entry]) => entry && isWarn(entry))
    .map(([name]) => name);
}

/**
 * Extract a list of all the rule names with the given prefix prepended to them
 * @param rules
 * @param prefix
 * @param shouldSort
 */
export function listRules(
  rules:
    | Record<string, Linter.RuleEntry | Rule.RuleModule>
    | Map<string, Linter.RuleEntry | Rule.RuleModule>
    | undefined,
  prefix?: string,
  shouldSort: boolean = false,
): string[] {
  const keys = rules instanceof Map ? [...rules.keys()] : Object.keys(rules ?? {});
  const names = keys.map((x) => `${prefix ? `${prefix}/` : ''}${x}`);
  return shouldSort ? names.sort((a, b) => a.localeCompare(b)) : names;
}

/**
 * Extract a list of rule names that are deprecated and, if provided, prepend the prefix to the rule name
 * @param rules
 * @param prefix
 */
export function getDeprecatedRules(
  rules: Record<string, Rule.RuleModule> | Map<string, Rule.RuleModule> | undefined,
  prefix: string = '',
): string[] {
  const entries = rules instanceof Map ? [...rules.entries()] : Object.entries(rules ?? {});
  return entries.filter(([, rule]) => rule.meta?.deprecated).map(([name]) => `${prefix ? `${prefix}/` : ''}${name}`);
}

interface Docs extends NonNullable<Rule.RuleMetaData['docs']> {
  requiresTypeChecking?: boolean;
}

/**
 * Extract a list of rule names that require type information, if provided, prepend the prefix to the rule name
 * @param rules
 * @param prefix
 */
export function getTypeInfoRequiredRules(
  rules: Record<string, Rule.RuleModule> | Map<string, Rule.RuleModule> | undefined,
  prefix: string = '',
): string[] {
  const entries = rules instanceof Map ? [...rules.entries()] : Object.entries(rules ?? {});
  return (
    entries
      // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
      .filter(([, rule]) => (rule.meta?.docs as Docs).requiresTypeChecking)
      .map(([name]) => `${prefix ? `${prefix}/` : ''}${name}`)
  );
}
