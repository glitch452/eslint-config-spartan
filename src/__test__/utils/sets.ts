type Widen<T> = T | (TSReset.WidenLiteral<T> & {});

/**
 * `A âˆ© B`.
 * Compute the intersection of unique elements contained in both {@link a} and {@link b}.
 * @param a
 * @param b
 * @returns
 */
export function intersection<A, B>(a: Widen<A>[], b: Widen<B>[]): Extract<Widen<A>, Widen<B>>[] {
  const setB = new Set<Widen<A> | Widen<B>>(b);
  return [...new Set<Widen<A>>(a)].filter((x): x is Extract<Widen<A>, Widen<B>> => setB.has(x));
}

/**
 * `A - B` a.k.a. `A / B`.
 * Compute the unique elements in {@link a} that do not also exist in {@link b}.
 * @param a
 * @param b
 * @returns
 */
export function difference<A, B>(a: A[], b: B[]): Widen<A>[] {
  const setB = new Set<A | B>(b);
  return [...new Set<A>(a)].filter((x): x is A => !setB.has(x));
}
