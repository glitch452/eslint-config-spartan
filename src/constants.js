export const configNamePrefix = 'eslint-config-spartan';

export const prefixes = /** @type {const} */ ({
  angular: '@angular-eslint',
  angularTemplate: '@angular-eslint/template',
  chaiExpect: 'chai-expect',
  chaiFriendly: 'chai-friendly',
  cypress: 'cypress',
  i18next: 'i18next',
  i18nJson: 'i18n-json',
  import: 'import',
  jest: 'jest',
  jestDom: 'jest-dom',
  jsDoc: 'jsdoc',
  jsxA11y: 'jsx-a11y',
  mdx: 'mdx',
  mocha: 'mocha',
  next: '@next/next',
  playwright: 'playwright',
  promise: 'promise',
  react: 'react',
  reactHooks: 'react-hooks',
  regExp: 'regexp',
  security: 'security',
  storybook: 'storybook',
  stylistic: '@stylistic',
  tailwindCss: 'tailwindcss',
  tanStackQuery: '@tanstack/query',
  testingLibrary: 'testing-library',
  typescriptEnum: 'typescript-enum',
  typescriptEsLint: '@typescript-eslint',
  unicorn: 'unicorn',
  unusedImports: 'unused-imports',
  vitest: 'vitest',
});
