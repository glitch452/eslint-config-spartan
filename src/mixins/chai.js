import chaiFriendlyPlugin from 'eslint-plugin-chai-friendly';
import chaiExpectPlugin from 'eslint-plugin-chai-expect';
import { configNamePrefix, prefixes } from '../constants.js';
import { files } from '../utils/index.js';
import { testEnvironmentAdjustments } from './testEnvironmentAdjustments.js';
/** @import { Linter } from 'eslint' */

/**
 * The `chai` mixin creates an ESLint config for
 * [eslint-plugin-chai-friendly](https://www.npmjs.com/package/eslint-plugin-chai-friendly) and
 * [eslint-plugin-chai-expect](https://www.npmjs.com/package/eslint-plugin-chai-expect) to assist with using the
 * [Chai Assertion Library](https://www.chaijs.com).
 *
 * Rule Prefixes:
 *
 * - The `eslint-plugin-chai-expect` rules are prefixed with `chai-expect`
 * - The `eslint-plugin-chai-friendly` rules are prefixed with `chai-friendly`
 *
 * Note: This config also adjusts some `@typescript-eslint` rules to make them more helpful for test files.
 * @param {object} options
 * @param {Linter.Config['files']=} options.files Set the files for this config. By default, this applies to all `js/ts` extension variants with `test.ext`, `spec.ext` and `.cy.ext`.
 * @returns {Linter.Config}
 */
export function chai(options = {}) {
  return {
    name: `${configNamePrefix}/${chai.name}`,
    files: options.files ?? [files.testSpec, files.cypress],
    plugins: {
      [prefixes.chaiFriendly]: chaiFriendlyPlugin,
      [prefixes.chaiExpect]: chaiExpectPlugin,
    },
    rules: {
      ...chaiFriendlyPlugin.configs.recommendedFlat.rules,
      ...chaiExpectPlugin.configs['recommended-flat'].rules,
      ...testEnvironmentAdjustments,
    },
  };
}
