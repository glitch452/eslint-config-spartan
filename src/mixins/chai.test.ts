import chaiFriendlyPlugin from 'eslint-plugin-chai-friendly';
import chaiExpectPlugin from 'eslint-plugin-chai-expect';
import typescriptEsLint from 'typescript-eslint';
import securityPlugin from 'eslint-plugin-security';
import unicornPlugin from 'eslint-plugin-unicorn';
import { Linter } from 'eslint';
import { chai } from './chai.js';
import { configSchema } from '../__test__/utils/configSchema.js';
import { difference, intersection } from '../__test__/utils/sets.js';
import { files } from '../utils/index.js';
import { CONFIG_NAME_PREFIX, prefixes } from '../constants.js';
import { getDeprecatedRules, getEnabledRules, getWarnRules, listRules } from '../__test__/utils/rules.js';

describe(chai.name, () => {
  const config = chai();

  const validRules = [
    ...new Linter({ configType: 'eslintrc' }).getRules().keys(),
    ...listRules(chaiFriendlyPlugin.rules, prefixes.chaiFriendly),
    ...listRules(chaiExpectPlugin.rules, prefixes.chaiExpect),
    ...listRules(securityPlugin.rules, prefixes.security),
    ...listRules(unicornPlugin.rules, prefixes.unicorn),
    ...listRules((typescriptEsLint.plugin as any).rules, prefixes.typescriptEsLint),
  ];
  const deprecatedRules = [
    ...getDeprecatedRules(new Linter({ configType: 'eslintrc' }).getRules()),
    ...getDeprecatedRules(chaiFriendlyPlugin.rules, prefixes.chaiFriendly),
    ...getDeprecatedRules(chaiExpectPlugin.rules, prefixes.chaiExpect),
    ...getDeprecatedRules(securityPlugin.rules, prefixes.security),
    ...getDeprecatedRules(unicornPlugin.rules, prefixes.unicorn),
    ...getDeprecatedRules((typescriptEsLint.plugin as any).rules, prefixes.typescriptEsLint),
  ];

  it('should create a valid eslint config', () => {
    const actual = () => configSchema.parse(config);
    expect(actual).not.toThrow();
  });

  it('should add the expected plugins to the config', () => {
    const actual = Object.keys(config.plugins ?? {}).sort();
    const expected = [prefixes.chaiExpect, prefixes.chaiFriendly].sort();
    expect(actual).toStrictEqual(expected);
  });

  it('should configure the expected rules', () => {
    expect(config.rules).toMatchSnapshot();
  });

  it('should set the name value', () => {
    expect(config.name).toBe(`${CONFIG_NAME_PREFIX}/${chai.name}`);
  });

  it('should set the default files value', () => {
    const actual = config.files;
    const expected = [files.testSpec, files.cypress];
    expect(actual).toStrictEqual(expected);
  });

  it('should set the given files value', () => {
    const filesValue = ['files.txt'];
    const actual = chai({ files: filesValue }).files;
    expect(actual).toStrictEqual(filesValue);
  });

  it('should only configure rules that exist', () => {
    const configuredRules = Object.keys(config.rules ?? {});
    const actual = difference(configuredRules, validRules);
    expect(actual).toStrictEqual([]);
  });

  it('should not configure rules that are deprecated', () => {
    const configuredAndEnabledRules = getEnabledRules(config.rules);
    const actual = intersection(configuredAndEnabledRules, deprecatedRules);
    expect(actual).toStrictEqual([]);
  });

  it('should not configure any rules as warn', () => {
    const configuredAndWarnRules = getWarnRules(config.rules);
    expect(configuredAndWarnRules).toStrictEqual([]);
  });
});
