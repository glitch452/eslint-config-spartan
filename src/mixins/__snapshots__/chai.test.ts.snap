// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`chai > should configure the expected rules 1`] = `
{
  "@typescript-eslint/consistent-type-assertions": [
    "error",
    {
      "assertionStyle": "as",
      "objectLiteralTypeAssertions": "allow",
    },
  ],
  "@typescript-eslint/no-explicit-any": "off",
  "@typescript-eslint/no-magic-numbers": "off",
  "@typescript-eslint/no-non-null-assertion": "off",
  "@typescript-eslint/no-unsafe-type-assertion": "off",
  "@typescript-eslint/no-unused-expressions": "off",
  "chai-expect/missing-assertion": "error",
  "chai-expect/no-inner-compare": "error",
  "chai-expect/no-inner-literal": "error",
  "chai-expect/terminating-properties": "error",
  "chai-friendly/no-unused-expressions": "error",
  "no-unused-expressions": "off",
  "security/detect-non-literal-fs-filename": "off",
  "unicorn/consistent-function-scoping": "off",
}
`;
